packages/react-dom-bindings/src/client/ReactDOMComponent.js

```diff
@@ -1188,15 +1198,415 @@ export function diffProperties(
   }
   if (styleUpdates) {
     if (__DEV__) {
-      validateShorthandPropertyCollisionInDev(styleUpdates, nextProps.style);
+      validateShorthandPropertyCollisionInDev(lastProps.style, nextProps.style);
     }
     (updatePayload = updatePayload || []).push('style', styleUpdates);
   }
   return updatePayload;
 }
 
-// Apply the diff.
 export function updateProperties(
+  domElement: Element,
+  tag: string,
+  lastProps: Object,
+  nextProps: Object,
+): void {
+  if (__DEV__) {
+    validatePropertiesInDevelopment(tag, nextProps);
+  }
+
+  switch (tag) {
+    case 'div':
+    case 'span':
+    case 'svg':
+    case 'path':
+    case 'a':
+    case 'g':
+    case 'p':
+    case 'li': {
+      // Fast track the most common tag types
+      break;
+    }
+    case 'input': {
+      // Update checked *before* name.
+      // In the middle of an update, it is possible to have multiple checked.
+      // When a checked radio tries to change name, browser makes another radio's checked false.
+      if (nextProps.type === 'radio' && nextProps.name != null) {
+        updateInputChecked(domElement, nextProps);
+      }
+      for (const propKey in lastProps) {
+        const lastProp = lastProps[propKey];
+        if (
+          lastProps.hasOwnProperty(propKey) &&
+          lastProp != null &&
+          !nextProps.hasOwnProperty(propKey)
+        ) {
+          switch (propKey) {
+            case 'checked': {
+              const checked = nextProps.defaultChecked;
+              const inputElement: HTMLInputElement = (domElement: any);
+              inputElement.checked =
+                !!checked &&
+                typeof checked !== 'function' &&
+                checked !== 'symbol';
+              break;
+            }
+            case 'value': {
+              // This is handled by updateWrapper below.
+              break;
+            }
+            // defaultChecked and defaultValue are ignored by setProp
+            default: {
+              setProp(domElement, tag, propKey, null, nextProps, lastProp);
+            }
+          }
+        }
+      }
+      for (const propKey in nextProps) {
+        const nextProp = nextProps[propKey];
+        const lastProp = lastProps[propKey];
+        if (
+          nextProps.hasOwnProperty(propKey) &&
+          nextProp !== lastProp &&
+          (nextProp != null || lastProp != null)
```

Comment:
(Why don't we cover `lastProp != null && nextProp == null` in the prior loop?)

Reply:
What do you mean?

(I change this in the upcoming PR anyway.)

Reply:
I think this case is basically just checking that we don't trigger an update if you transition from `undefined` to `null` or vice versa. Seems unnecessary tbh.

Reply:
I think this case is basically just checking that we don't trigger an update if you transition from `undefined` to `null` or vice versa. Seems unnecessary tbh.

Reply:
Most props have some non-null value that's the same as deleted now (like function/symbol) so it's not really much extra work also check for null there but maybe. `setProp` still needs to deal with it. I considered having a separate `deleteProp` instead of setting null but because it has to go through the special case route and each of those have some other way to delete already, it didn't make it worth it.